tests/validate_test.py

```diff
@@ -14,107 +14,262 @@
+            e = cm.exception
+            self.assertEqual(str(e), f"invalid {msg.DESCRIPTOR.name}")
+            self._compare_violations(e.violations, [expected_violation])
+
+        # Test collect_violations
+        violations = validator.collect_violations(msg)
+        self._compare_violations(violations, [expected_violation])
+
+    def _run_valid_tests(self, msg: message.Message):
+        """A helper function for testing successful validation on a given message
+
+        The tests are run using validators created via all possible methods and
+        validation is done via a call to `validate` as well as a call to `collect_violations`.
+        """
+        for label, v in get_default_validator():
+            with self.subTest(label=label):
+                # Test validate
+                try:
+                    v.validate(msg)
+                except Exception:
+                    self.fail(f"[{label}]: unexpected validation failure")
+
+                # Test collect_violations
+                violations = v.collect_violations(msg)
+                self.assertEqual(len(violations), 0)
+
+    def _run_invalid_tests(self, msg: message.Message, expected: list[rules.Violation]):
+        """A helper function for testing unsuccessful validation on a given message
+
+        The tests are run using validators created via all possible methods and
+        validation is done via a call to `validate` as well as a call to `collect_violations`.
+        """
+        for label, v in get_default_validator():
+            with self.subTest(label=label):
+                # Test validate
+                with self.assertRaises(protovalidate.ValidationError) as cm:
+                    v.validate(msg)
+                e = cm.exception
+                self.assertEqual(str(e), f"invalid {msg.DESCRIPTOR.name}")
+                self._compare_violations(e.violations, expected)
+
+                # Test collect_violations
+                violations = v.collect_violations(msg)
+                self._compare_violations(violations, expected)
+
+    def _run_compilation_error_tests(self, msg: message.Message, expected: str):
+        """A helper function for testing compilation errors when validating.
+
+        The tests are run using validators created via all possible methods and
+        validation is done via a call to `validate` as well as a call to `collect_violations`.
+        """
+        for label, v in get_default_validator():
+            with self.subTest(label=label):
+                # Test validate
+                with self.assertRaises(protovalidate.CompilationError) as vce:
+                    v.validate(msg)
+                    assert str(vce.exception) == expected
+
+                # Test collect_violations
+                with self.assertRaises(protovalidate.CompilationError) as cvce:
+                    v.collect_violations(msg)
+                    assert str(cvce.exception) == expected
```

Comment:
nit: use `self.assertEqual` to compare the exception strings? we're nearly fully on the `unittest` style of assertions, even though we're using the `pytest` runner, so makes sense to be consistent. 🤷